Cannot compile locally with Python 3.12

### Bug Description

With the release of Python 3.12, I've tried to install my project in a 3.12 venv. I get the following error:

```
(venv3.12) lily@lily:~/work/kloppindustries/kolo/python$ maturin develop --extras=test_generation
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings
🐍 Found CPython 3.12d at /home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python
📡 Using build options profile, features from pyproject.toml
/home/lily/work/kloppindustries/kolo/python/venv3.12/lib/python3.12/site-packages/pip/_internal/metadata/importlib/_envs.py:111: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pip._vendor.pkg_resources import find_distributions
Ignoring tomli: markers 'python_version < "3.11"' don't match your environment
Requirement already satisfied: asgiref>=3.6.0 in ./venv3.12/lib/python3.12/site-packages (3.7.2)
Requirement already satisfied: cerberus>=1.3.4 in ./venv3.12/lib/python3.12/site-packages (1.3.5)
Requirement already satisfied: click>=7.1.2 in ./venv3.12/lib/python3.12/site-packages (8.1.7)
Requirement already satisfied: httpx in ./venv3.12/lib/python3.12/site-packages (0.25.0)
Requirement already satisfied: simplejson>=3.8.0 in ./venv3.12/lib/python3.12/site-packages (3.19.1)
Requirement already satisfied: toolz>=0.11.1 in ./venv3.12/lib/python3.12/site-packages (0.12.0)
Requirement already satisfied: ulid-py>=1.1.0 in ./venv3.12/lib/python3.12/site-packages (1.1.0)
Requirement already satisfied: jinja2>=3.0.0 in ./venv3.12/lib/python3.12/site-packages (3.1.2)
Requirement already satisfied: more_itertools in ./venv3.12/lib/python3.12/site-packages (10.1.0)
Requirement already satisfied: sqlglot>=16.1.1 in ./venv3.12/lib/python3.12/site-packages (18.10.1)
Requirement already satisfied: certifi in ./venv3.12/lib/python3.12/site-packages (from httpx) (2023.7.22)
Requirement already satisfied: httpcore<0.19.0,>=0.18.0 in ./venv3.12/lib/python3.12/site-packages (from httpx) (0.18.0)
Requirement already satisfied: idna in ./venv3.12/lib/python3.12/site-packages (from httpx) (3.4)
Requirement already satisfied: sniffio in ./venv3.12/lib/python3.12/site-packages (from httpx) (1.3.0)
Requirement already satisfied: MarkupSafe>=2.0 in ./venv3.12/lib/python3.12/site-packages (from jinja2>=3.0.0) (2.1.3)
Requirement already satisfied: anyio<5.0,>=3.0 in ./venv3.12/lib/python3.12/site-packages (from httpcore<0.19.0,>=0.18.0->httpx) (4.0.0)
Requirement already satisfied: h11<0.15,>=0.13 in ./venv3.12/lib/python3.12/site-packages (from httpcore<0.19.0,>=0.18.0->httpx) (0.14.0)
   Compiling pyo3-ffi v0.19.2
error[E0425]: cannot find function `_Py_NegativeRefcount` in this scope
   --> /home/lily/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-ffi-0.19.2/src/object.rs:586:17
    |
457 |     pub fn _Py_NegativeRefCount(filename: *const c_char, lineno: c_int, op: *mut PyObject);
    |     --------------------------------------------------------------------------------------- similarly named function `_Py_NegativeRefCount` defined here
...
586 |                 _Py_NegativeRefcount(location.file(), location.line(), op);
    |                 ^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists (notice the capitalization): `_Py_NegativeRefCount`

For more information about this error, try `rustc --explain E0425`.
error: could not compile `pyo3-ffi` (lib) due to previous error
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `PYO3_ENVIRONMENT_SIGNATURE="cpython-3.12-64bit" PYO3_PYTHON="/home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python" PYTHON_SYS_EXECUTABLE="/home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python" "cargo" "rustc" "--profile" "release" "--features" "pyo3/extension-module" "--message-format" "json-render-diagnostics" "--manifest-path" "/home/lily/work/kloppindustries/kolo/python/rust/Cargo.toml" "--lib"`
```

I don't see this error when I use Python 3.11 locally. I also don't see it for any of our CI builds, including for Python 3.12.

I installed Python 3.12 using pyenv.

### Your maturin version (`maturin --version`)

maturin 1.3.0

### Your Python version (`python -V`)

Python 3.12.0

### Your pip version (`pip -V`)

pip 23.2.1

### What bindings you're using

pyo3

### Does `cargo build` work?

- [ ] Yes, it works

### If on windows, have you checked that you aren't accidentally using unix path (those with the forward slash `/`)?

- [ ] Yes

### Steps to Reproduce

1. Install Python 3.12.0 with pyenv
2. Create a venv:

```
lily@lily:~/work/kloppindustries/kolo/python$ python3.12 -m venv venv3.12
lily@lily:~/work/kloppindustries/kolo/python$ source venv3.12/bin/activate
```

3. Try to build:

```
(venv3.12) lily@lily:~/work/kloppindustries/kolo/python$ RUST_LOG=maturin=debug maturin build
2023-10-04T13:51:49.088137Z DEBUG maturin::project_layout: Found pyproject.toml in working directory at "/home/lily/work/kloppindustries/kolo/python/pyproject.toml"
2023-10-04T13:51:49.090140Z DEBUG maturin::project_layout: Python first src-layout detected
2023-10-04T13:51:49.090452Z DEBUG maturin::project_layout: Resolving cargo metadata from "/home/lily/work/kloppindustries/kolo/python/rust/Cargo.toml"
2023-10-04T13:51:49.140934Z DEBUG maturin::project_layout: Project layout resolved project_root=/home/lily/work/kloppindustries/kolo/python python_dir=/home/lily/work/kloppindustries/kolo/python/src rust_module=/home/lily/work/kloppindustries/kolo/python/src/kolo python_module=/home/lily/work/kloppindustries/kolo/python/src/kolo extension_name=_kolo module_name=kolo._kolo
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings
2023-10-04T13:51:49.244900Z DEBUG maturin::python_interpreter: Found CPython interpreter at /home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python
🐍 Found CPython 3.12d at /home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python
📡 Using build options profile, features from pyproject.toml
2023-10-04T13:51:49.245073Z DEBUG maturin::compile: Setting PYO3_PYTHON to /home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python
2023-10-04T13:51:49.245082Z DEBUG maturin::compile: Running PYO3_ENVIRONMENT_SIGNATURE="cpython-3.12-64bit" PYO3_PYTHON="/home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python" PYTHON_SYS_EXECUTABLE="/home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python" "cargo" "rustc" "--profile" "release" "--features" "pyo3/extension-module" "--message-format" "json-render-diagnostics" "--manifest-path" "/home/lily/work/kloppindustries/kolo/python/rust/Cargo.toml" "--lib"
   Compiling pyo3-build-config v0.19.2
   Compiling pyo3-ffi v0.19.2
   Compiling pyo3 v0.19.2
   Compiling _kolo v2.15.1 (/home/lily/work/kloppindustries/kolo/python/rust)
error[E0425]: cannot find function `_Py_NegativeRefcount` in this scope
   --> /home/lily/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-ffi-0.19.2/src/object.rs:586:17
    |
457 |     pub fn _Py_NegativeRefCount(filename: *const c_char, lineno: c_int, op: *mut PyObject);
    |     --------------------------------------------------------------------------------------- similarly named function `_Py_NegativeRefCount` defined here
...
586 |                 _Py_NegativeRefcount(location.file(), location.line(), op);
    |                 ^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists (notice the capitalization): `_Py_NegativeRefCount`

For more information about this error, try `rustc --explain E0425`.
error: could not compile `pyo3-ffi` (lib) due to previous error
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `PYO3_ENVIRONMENT_SIGNATURE="cpython-3.12-64bit" PYO3_PYTHON="/home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python" PYTHON_SYS_EXECUTABLE="/home/lily/work/kloppindustries/kolo/python/venv3.12/bin/python" "cargo" "rustc" "--profile" "release" "--features" "pyo3/extension-module" "--message-format" "json-render-diagnostics" "--manifest-path" "/home/lily/work/kloppindustries/kolo/python/rust/Cargo.toml" "--lib"
```

I don't have a minimal reproducing example yet.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -376,8 +376,8 @@ jobs:
           components: rust-src
       - name: Install python3 standalone debug build with nox
         run: |
-          PBS_RELEASE="20230826"
-          PBS_PYTHON_VERSION="3.11.5"
+          PBS_RELEASE="20231002"
+          PBS_PYTHON_VERSION="3.12.0"
           PBS_ARCHIVE="cpython-${PBS_PYTHON_VERSION}+${PBS_RELEASE}-x86_64-unknown-linux-gnu-debug-full.tar.zst"
           wget "https://github.com/indygreg/python-build-standalone/releases/download/${PBS_RELEASE}/${PBS_ARCHIVE}"
           tar -I zstd -xf "${PBS_ARCHIVE}"
@@ -393,10 +393,10 @@ jobs:
           PYO3_CONFIG_FILE=$(mktemp)
           cat > $PYO3_CONFIG_FILE << EOF
           implementation=CPython
-          version=3.11
+          version=3.12
           shared=true
           abi3=false
-          lib_name=python3.11d
+          lib_name=python3.12d
           lib_dir=${{ github.workspace }}/python/install/lib
           executable=${{ github.workspace }}/python/install/bin/python3
           pointer_width=64

--- a/pyo3-ffi/src/object.rs
+++ b/pyo3-ffi/src/object.rs
@@ -452,18 +452,14 @@ pub const Py_TPFLAGS_DEFAULT: c_ulong = if cfg!(Py_3_10) {
 pub const Py_TPFLAGS_HAVE_FINALIZE: c_ulong = 1;
 pub const Py_TPFLAGS_HAVE_VERSION_TAG: c_ulong = 1 << 18;
 
-#[cfg(all(py_sys_config = "Py_REF_DEBUG", not(Py_LIMITED_API)))]
 extern "C" {
-    pub fn _Py_NegativeRefCount(filename: *const c_char, lineno: c_int, op: *mut PyObject);
-    #[cfg(Py_3_12)]
-    #[link_name = "_Py_IncRefTotal_DO_NOT_USE_THIS"]
-    fn _Py_INC_REFTOTAL();
-    #[cfg(Py_3_12)]
-    #[link_name = "_Py_DecRefTotal_DO_NOT_USE_THIS"]
-    fn _Py_DEC_REFTOTAL();
-}
+    #[cfg(all(py_sys_config = "Py_REF_DEBUG", not(Py_LIMITED_API)))]
+    pub fn _Py_NegativeRefcount(filename: *const c_char, lineno: c_int, op: *mut PyObject);
+    #[cfg(all(Py_3_12, py_sys_config = "Py_REF_DEBUG", not(Py_LIMITED_API)))]
+    fn _Py_INCREF_IncRefTotal();
+    #[cfg(all(Py_3_12, py_sys_config = "Py_REF_DEBUG", not(Py_LIMITED_API)))]
+    fn _Py_DECREF_DecRefTotal();
 
-extern "C" {
     #[cfg_attr(PyPy, link_name = "_PyPy_Dealloc")]
     pub fn _Py_Dealloc(arg1: *mut PyObject);
 
@@ -537,7 +533,7 @@ pub unsafe fn Py_INCREF(op: *mut PyObject) {
         // or submit a PR supporting Py_STATS build option and pystats.h
 
         #[cfg(all(py_sys_config = "Py_REF_DEBUG", Py_3_12))]
-        _Py_INC_REFTOTAL();
+        _Py_INCREF_IncRefTotal();
     }
 }
 
@@ -584,7 +580,7 @@ pub unsafe fn Py_DECREF(op: *mut PyObject) {
         // or submit a PR supporting Py_STATS build option and pystats.h
 
         #[cfg(all(py_sys_config = "Py_REF_DEBUG", Py_3_12))]
-        _Py_DEC_REFTOTAL();
+        _Py_DECREF_DecRefTotal();
 
         #[cfg(Py_3_12)]
         {
@@ -593,7 +589,8 @@ pub unsafe fn Py_DECREF(op: *mut PyObject) {
             #[cfg(py_sys_config = "Py_REF_DEBUG")]
             if (*op).ob_refcnt.ob_refcnt < 0 {
                 let location = std::panic::Location::caller();
-                _Py_NegativeRefcount(location.file(), location.line(), op);
+                let filename = std::ffi::CString::new(location.file()).unwrap();
+                _Py_NegativeRefcount(filename.as_ptr(), location.line() as i32, op);
             }
 
             if (*op).ob_refcnt.ob_refcnt == 0 {

